runtime/complex.go

```diff
@@ -169,6 +178,169 @@ func complexCoerce(o *Object) (complex128, bool) {
 	return complex(floatO, 0.0), true
 }
 
+func complexNew(f *Frame, t *Type, args Args, _ KWArgs) (*Object, *BaseException) {
+	argc := len(args)
+	if argc == 0 {
+		return newObject(t), nil
+	}
+	if argc > 2 {
+		return nil, f.RaiseType(TypeErrorType, "'__new__' of 'complex' requires at most 2 arguments")
+	}
+	if t != ComplexType {
+		// Allocate a plain complex then copy it's value into an object
+		// of the complex subtype.
+		x, raised := complexNew(f, ComplexType, args, nil)
+		if raised != nil {
+			return nil, raised
+		}
+		result := toComplexUnsafe(newObject(t))
+		result.value = toComplexUnsafe(x).Value()
+		return result.ToObject(), nil
+	}
+	if complexSlot := args[0].typ.slots.Complex; complexSlot != nil && argc == 1 {
+		c, raised := complexConvert(complexSlot, f, args[0])
+		if raised != nil {
+			return nil, raised
+		}
+		return c.ToObject(), nil
+	}
+	if args[0].isInstance(StrType) {
+		if argc > 1 {
+			return nil, f.RaiseType(TypeErrorType, "complex() can't take second arg if first is a string")
+		}
+		s := toStrUnsafe(args[0]).Value()
+		result, err := parseComplex(s)
+		if err != nil {
+			return nil, f.RaiseType(ValueErrorType, "complex() arg is a malformed string")
+		}
+		return NewComplex(result).ToObject(), nil
+	}
+	if argc > 1 && args[1].isInstance(StrType) {
+		return nil, f.RaiseType(TypeErrorType, "complex() second arg can't be a string")
+	}
+	cr, raised := complex128Convert(f, args[0])
+	if raised != nil {
+		return nil, raised
+	}
+	var ci complex128
+	if argc > 1 {
+		ci, raised = complex128Convert(f, args[1])
+		if raised != nil {
+			return nil, raised
+		}
+	}
+	return NewComplex(cr + ci*1i).ToObject(), nil
+}
+
+func complex128Convert(f *Frame, o *Object) (complex128, *BaseException) {
+	if complexSlot := o.typ.slots.Complex; complexSlot != nil {
+		c, raised := complexConvert(complexSlot, f, o)
+		if raised != nil {
+			return complex(0, 0), raised
+		}
+		return c.Value(), nil
+	} else if floatSlot := o.typ.slots.Float; floatSlot != nil {
+		result, raised := floatConvert(floatSlot, f, o)
+		if raised != nil {
+			return complex(0, 0), raised
+		}
+		return complex(result.Value(), 0), nil
+	} else {
+		return complex(0, 0), f.RaiseType(TypeErrorType, "complex() argument must be a string or a number")
+	}
+}
+
+func complexConvert(complexSlot *unaryOpSlot, f *Frame, o *Object) (*Complex, *BaseException) {
```

Comment:
Please move helper functions to the end of the file. The general order of stuff in these files is: type definition, methods, type init function, helpers.